DOC: added Docker quickstart guide by @andyfaff

mdhaber opened the pull request:
Closes gh-8340.
Converted @andyfaff's MD [instructions](https://github.com/scipy/scipy/wiki/Development-Environment-Quickstart-Guide-(Docker)-(WIP)) to RST and added to the project.

mdhaber commented:
A few questions:

- [ ] I don't quite understand the part of the `docker run` command `$PWD/scipy:/home/scipy`. `$PWD` should be the root directory of the SciPy repository on the computer, which would make `$PWD/scipy` the `scipy/scipy` directory, right? But `/home/scipy` in the container should be the root `scipy` directory so we can see `setup.py`, etc..., not `scipy/scipy`. 
- [ ] My last post in gh-8340 was about my directory running out of space during build. Do you know how to fix this?
- [ ] Is it possible to start the container and automatically run the commands in steps 4-6?

andyfaff commented:
>  My last post in gh-8340 was about my directory running out of space during build. Do you know how to fix this?

I'm not sure about this, because I've not experienced it. In the Docker preferences you can change:

1) the size of Swap (I don't think that would be related though).
2) the Disk Image Size. I think this is the cumulative size of all the images. Mines currently set to 64 Gb, with 30.5 Gb taken up. If this is too small then that might cause problems. 
It may be related to the containers file system persisting after the container process stops. See the [`--rm` flag for details](https://docs.docker.com/engine/reference/run/#clean-up---rm). 

> By default a container’s file system persists even after the container exits. This makes debugging a lot easier (since you can inspect the final state) and you retain all your data by default. But if you are running short-term foreground processes, these container file systems can really pile up.  If instead you’d like Docker to automatically clean up the container and remove the file system when the container exits, you can add the --rm flag

> Is it possible to start the container and automatically run the commands in steps 4-6?

I can't tell which steps you mean here. It is probable that steps could be run when a container is started up, it's Linux after all, but I don't know how to. Plus, the steps would have to be transparent to the user. I think it would be preferred that the user initiate things that will access resources outside the container, such as downloads.

> I don't quite understand the part of the docker run command $PWD/scipy:/home/scipy. $PWD should be the root directory of the SciPy repository on the computer, which would make $PWD/scipy the scipy/scipy directory, right? But /home/scipy in the container should be the root scipy directory so we can see setup.py, etc..., not scipy/scipy. (Yet everything works, so I think just the explanation needs to be tweaked, not the command itself.)

This part of the docker run command is `-v USERDIR:CONTAINERDIR`. `USERDIR` is a folder/volume on the host computer specified by an absolute path that you want to make available on the container. `CONTAINERDIR` is the absolute path that the folder will have on the container.
In our case we want to make USERDIR the directory where the setup.py file is. If you're already in the directory that holds setup.py then we'd want `-v $PWD:/home/scipy`, if we were in the directory above that then we'd want to use `-v $PWD/scipy:/home/scipy`.

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+   Note that this creates in the terminal’s working directory a
+   ``scipy`` folder containing the SciPy source code. This assumes that
+   you have a ``git`` command line client that is available on your
+   PATH; if not, you can follow these `instructions to install a git client`_.
```
There is a git client in the docker image, so an alternative could be `git clone` in the container. That might make subsequent steps more complicated.

mdhaber replied:
Yeah. I think installing git in host OS is OK. If they didn't, they'd still have to create a folder that can be shared between the host OS and docker (if they want access to the SciPy files from the host OS), right?

andyfaff replied:
>  If they didn't, they'd still have to create or choose a folder that can be shared between the host OS and docker (if they want access to the SciPy files from the host OS), right?

Yes.

If they don't install git on the host then they would have to fire up the container everytime they wanted to use git to push/pull from github..

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+container do not persist after you close it.
+
+#. In a terminal window, change the directory (using the ``cd`` command)
+   to the folder that contains the SciPy git repository.
```
> to the root folder of the SciPy git repository (holding the main `setup.py` file)

mdhaber replied:
OK, that is probably more clear. I just copy pasted from wiki.

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+   SciPy Docker container by entering the following command in a terminal
+   window::
+
+      docker run -it --rm -v $PWD/scipy:/home/scipy scipy/scipy-dev /bin/bash
```
`docker run -it --rm -v $PWD:/home/scipy scipy/scipy-dev /bin/bash`

mdhaber replied:
This was copy-pasted from the wiki instructions.
And it's what I asked about above:
> I don't quite understand the part of the docker run command $PWD/scipy:/home/scipy.

That makes much more sense. 
(Strange that `docker run -it --rm -v $PWD/scipy:/home/scipy scipy/scipy-dev /bin/bash` worked, though...)

andyfaff replied:
your way, being in the root folder of the repository does make more sense.

```
docker run -it --rm -v $PWD:/home/scipy scipy/scipy-dev /bin/bash
```

mdhaber replied:
OK I'll change.

mdhaber replied:
That didn't work on my computer, but adding `/` after `PWD` did:
```
docker run -it --rm -v $PWD/:/home/scipy scipy/scipy-dev /bin/bash
```

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+   named ``scipy-dev`` (based on Ubuntu Bionic) from the ``scipy``
+   `Docker Hub repository`_. When the Docker container starts, the
+   ``scipy`` directory from the current directory of the host (``$PWD``) is
+   made available in the container as ``/home/scipy``. The changes you make
```
When the Docker container starts, the current directory of the host (``$PWD``) is made available in the container as ``/home/scipy``. The changes you make from the container to any of the files in that directory are also visible in the host, and vice-versa.

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+   suite of tests that make sure SciPy is working as it should, and ``-v``
+   activates the ``–verbose`` option to show all the test output.
+
+From here, you can start a Python console (e.g. enter ``python3.7``) or
```
Is this the best way of doing this? I normally do a `pip install .`, navigate to a different directory and start using Python.

mdhaber replied:
If you install that way, don't you have to install again after making changes to the source code?

andyfaff replied:
> Is what the best way?

Of having the changes you just made to scipy code available in a newly opened interpreter.

> Do you have to do pip install . every time you change the source code?

If I'm not using runtests.py that's what I currently do. But if there are easier ways I'm all for them.

mdhaber replied:
I think `pip install -e .` makes symlinks instead of copying files so you don't have to run it after making changes to source for changes to take effect.

mdhaber replied:
I haven't used that before, though. I should try it out.

Adding the folder to PYTHONPATH, `conda develop`, and `python setup.py develop` have all made it so I don't have to repeat the install, but adding to PYTHONPATH was the first I learned so I guess I stuck with that.

mdhaber replied:
@rgommers do you have a recommendation for which of the options we should add to these instructions?

- `pip install -e .`
- `python setup.py develop`
- manually add SciPy root to PYTHONPATH

Note that in the other quickstarts, `conda develop` is used. I think that is appropriate there because `conda` is used to install everything and manage virtual environments, but this tutorial doesn't use Anaconda.

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+container as you would any other Linux terminal. You've already cloned
+SciPy on your computer, and git and all required compilers are already
+installed, so you can simply skip the corresponding steps
+(:ref:`quickstart-ubuntu-build` 2-7, 9, 10, and 13).
```
fixed point numbers are going to go stale quickly.

mdhaber replied:
I know : / I debated. I wanted to make it simple, but I guess for the sake of maintenance it is better to leave them out.

mdhaber replied:
Let me unresolve until I make the change.

mdhaber commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,177 @@
+ system tools, system libraries and settings.
+ Container images become containers at runtime and in the case of Docker
+ containers - images become containers when they run on Docker Engine.
+ Available for both Linux and Windows-based applications, containerized
```
I think what it means is that Docker Containers can be Windows-based, not just Linux-based. https://hub.docker.com/_/microsoft-windows-base-os-images

andyfaff replied:
That's nice to know.

mdhaber commented:
@andyfaff I just tried building the docs but it fails if I have only set up SciPy for use with Python 3.7 because in the doc Makefile:
`python3 -c 'from sys import version_info as v; print("{0}.{1}".format(v[0], v[1]))'`
evaluates to 3.6.

I set up SciPy for use with 3.6 and everything worked. 
Any thoughts on how this should be addressed? The easiest would be to say that the docs can only be built with 3.6, or to manually modify the Makefile, but can you think of something better?

andyfaff commented:
I'm not sure what the requirements for building the documentation are. I would assume sphinx, matplotlib, numpydoc, possibly others. Documentation can be built by using:
```
python setup.py build_sphinx
```
This is the preferred way according to `doc/README.txt`. On macOS the documentation build works with Python3.7.

So text for the document could be:

```
Documentation can be built by using the `python3.7 setup.py build_sphinx` command, and will require the `sphinx, matplotlib, numpydoc` packages to be installed.
```

mdhaber commented:
@andyfaff the additional requirements are `matplotlib` and `sphinx`. That's why I asked about `matplotlib`. I didn't ask about `sphinx` because the documentation for building docs in the contributor guide specifies that `sphinx` is also needed.

I will see if that way of building the docs makes a difference.

andyfaff commented:
> I didn't ask about sphinx because the documentation for building docs in the contributor guide specifies that sphinx is also needed.

Perhaps that location needs to have matplotlib added as well.

mdhaber commented:
The other quickstarts guides already install `matplotlib`, but maybe you're right. I could take it out of those and add it to the doc rendering guide.

andyfaff commented on doc/source/dev/contributor/quickstart_docker.rst:
```diff
@@ -0,0 +1,176 @@
+   suite of tests that make sure SciPy is working as it should, and ``-v``
+   activates the ``–verbose`` option to show all the test output.
+
+#. (Optional) If you'll want to use SciPy from any directory other than the
```
If *you*

mdhaber replied:
If you prefer, I can change it, but this was a contraction if "If you will want to...", which is fine, right?

andyfaff replied:
Much of a muchness

andyfaff commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -21,9 +21,9 @@ Rendering Documentation Locally
+#. Install `Sphinx`_ and `matplotlib`_. For
    example, if you're using the Anaconda distribution of Python, enter in a
-   terminal window ``conda install sphinx`` or ``conda update sphinx``.
+   terminal window ``conda install sphinx matplotlib``.
```
you need numpydoc as well.

mdhaber replied:
Somehow I think it gets installed without listing explicitly, because I built the docs I'm the container (with `make html-scipyorg`) without explicitly installing it.

mdhaber replied:
I think it's managed with `git submodule`

rgommers replied:
Indeed:
```
$ cat .gitmodules 
[submodule "doc/scipy-sphinx-theme"]
	path = doc/scipy-sphinx-theme
	url = https://github.com/scipy/scipy-sphinx-theme.git
[submodule "doc/sphinxext"]
	path = doc/sphinxext
	url = https://github.com/numpy/numpydoc.git
[submodule "doc/source/_static/scipy-mathjax"]
	path = doc/source/_static/scipy-mathjax
	url = https://github.com/scipy/scipy-mathjax.git
```

when you clone the repo and want to build the docs, you should do. I believe you get a message to that extent if you forget.
```
git submodule init
git submodule update
```

mdhaber replied:
Thanks. In gh-10724, @peterbell10 suggested simplifying:
`git submodule update --init`

andyfaff commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -21,9 +21,9 @@ Rendering Documentation Locally
    example, if you're using the Anaconda distribution of Python, enter in a
-   terminal window ``conda install sphinx`` or ``conda update sphinx``.
+   terminal window ``conda install sphinx matplotlib``.
 #. In a terminal window, browse to the ``scipy/doc`` directory. Note the
```
As already discussed I think `python setup.py build_sphinx` is the recommended way of building doc.

mdhaber replied:
I still need to try that in the container.
I did try it on my Mac, but it failed.

mdhaber replied:
And in the docker container it doesn't build the SciPy documentation, it builds some sort of `scipy-sphinx-theme` documentation:
![image](https://user-images.githubusercontent.com/6570539/63551820-f8f97480-c4ea-11e9-9f9d-235e0d523eea.png)

mdhaber replied:
The instructions [here](https://14871-1460385-gh.circle-artifacts.com/0/html-scipyorg/dev/contributor/rendering_documentation.html#rendering-documentation) work fine as long as I set up the development environment using Python 3.6. Alternatively, if I manually specify `PYVER:=3.7` in the Makefile, rendering the documentation with Python 3.7 works fine. 

Do you develop with Python 3.6? I'm guess that new developers who need this guide will use 3.7, so could we remove 3.6 from the container? I could add instructions on changing the Makefile; but that might not be the best approach either.

andyfaff replied:
I develop with 3.7. There's no point removing 3.6 because that's the default Python for Ubuntu:bionic.

I've just build the documentation using `python setup.py build_sphinx` in docker. The documentation show up in `scipy/build/sphinx/html/`. The index.html looks a little funny though:
![Untitled](https://user-images.githubusercontent.com/702934/63561363-096a2780-c59d-11e9-9278-cbf689763766.png)

mdhaber replied:
Take a look at whether it produced output in html-scipyorg.

andyfaff replied:
I couldn't find an html-scipyorg folder.

andyfaff commented:
When I try to build using the makefile I get:
```
root@0e7bb8cf74a5:/home/scipy/doc# make html
mkdir -p build/html build/doctrees
LANG=C python3.6 /usr/local/bin/sphinx-build -b html -d build/doctrees   source build/html 
Traceback (most recent call last):
  File "/usr/local/bin/sphinx-build", line 6, in <module>
    from sphinx.cmd.build import main
ModuleNotFoundError: No module named 'sphinx'
Makefile:109: recipe for target 'html' failed
make: *** [html] Error 1
```
If I go in and amend PYVER in the Makefile to call python3.7 then I don't get much further:
```
root@0e7bb8cf74a5:/home/scipy/doc# make html
mkdir -p build/html build/doctrees
LANG=C python3.7 /usr/local/bin/sphinx-build -b html -d build/doctrees   source build/html 
Running Sphinx v2.2.0

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.7/dist-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/home/scipy/doc/source/conf.py", line 23, in <module>
    import scipy._lib.uarray as ua
ModuleNotFoundError: No module named 'scipy'

Makefile:109: recipe for target 'html' failed
make: *** [html] Error 2
```
This looks like scipy has to be installed before it'll work. So I then tried doing a `python3.7 setup.py develop` and a `make html`, and this time the doc started building. It's worth making clear in the documentation that scipy has to be importable (e.g. via installation) before `make html` will work. Oh, and that the Makefile may have to be adjusted for the Python version you're using, `python3` in the docker container resolves to python3.6.8.

mdhaber commented:
> It's worth making clear in the documentation that scipy has to be importable

Step 0 in the documentation rendering guide was recently added to address that, but I'll make it even more explicit. And I'll remove the "optional" label from the `python setup.py develop` step in this guide.  The other quickstarts have equivalent `conda develop` commands, so as long as the user follows a quickstart, they should be all set.

> Oh, and that the Makefile may have to be adjusted 

Yes, or it sounds like @rgommers had an idea in gh-10708 for how the Makefile can be adjusted to work even if the user has multiple versions installed. I'll wait to see if we can do that. I suspect that there would be a lot of PRs with accidental Makefile changes if we suggest Makefile changes in the documentation ; )

mdhaber commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -34,8 +37,7 @@ To render the documentation on your own machine (macOS or Linux):
    to execute the documentation build instructions from the ``Makefile``.
    This can take a while the first time, but subsequent documentation builds
-   are typically much faster. *Note: If you use a virtual environment for
-   development, activate it first.*
```
This needs to be done before installing Sphinx and matplotlib, and step 0 now states "Ensure that you have a working SciPy :ref:`dev-env` active", so I think this should be removed.

If we want to make things like this more explicit (e.g. add instructions for testing whether `import scipy; scipy.__version__` shows the development version number from any arbitrary location), we should add a troubleshooting section. I think this is a good idea for this and other contributor guides. That way the main text can be concise, providing a rather minimal set of instructions that should work for everyone if followed carefully, and common issues can be listed separately for those who run into trouble.

mdhaber commented on doc/source/dev/contributor/rendering_documentation.rst:
```diff
@@ -57,12 +59,12 @@ on the cloud.
 #. Back in GitHub, at the bottom of the PR, select “Show all Checks”.
-#. Next to “ci/circleci: build_docs”, select “Details”.
-#. Select the “Artifacts” tab and find ``index.html``.
+#. Next to “ci/circleci: build_docs artifact”, select “Details”.
```
Well, I thought this is how it's supposed to work with gh-10706, but it's not looking like that for me yet.

mdhaber replied:
Update: It will if PR is rebased on master since gh-10706 merged.

mdhaber commented:
Per the discussion in gh-10715, should the different Python versions be in separate virtual environments? This would solve the issue of `make html-scipyorg` choosing the wrong Python.

rgommers commented:
> Per the discussion in gh-10715, should the different Python versions be in separate virtual environments? 

Sorry, that part of my comment was a bit too short. for `conda` that makes sense, but a `virtualenv` cannot change/manage Python version. The question is where this is coming from. Using a non-default Python install outside of `conda` seems unhealthy. If this situation occurs, for example on a Linux system where you can have system Python and a user Python, the user Python should have preference so you can simply refer to it as `python`.

mdhaber commented:
Thanks. I'm only familiar with `conda`.

This image we're working with, which I believe comes from the Dockerfile in gh-10655, has (I think - please correct me, @andyfaff) the Ubuntu default Python 3.6 and an apt-get installed Python 3.7 along with corresponding pips. No conda.

mdhaber commented:
Revise doc build instructions according to gh-10724 before merging.

andyfaff commented:
The error was a shared library linked against `libmkl_rt.so` not being able to find the shared library at runtime. This happens if you build scipy using a Python environment with numpy from conda. If you then try to run code that depends on that shared library, then you'll get the error.
You could easily get this if you build on the host, then in docker.

mdhaber commented:
I think this is about ready. I wanted to clean up the commit history a bit with:
`git rebase -i c13fc37924e8e1a2533787ee22453363b8dc17e2`
but the list of commits looks a lot messier than I expected, probably because there were some merges. 
```
pick 15a441895 DOC: revised Docker quickstart guide
pick 3f8ddd206 DOC: added note about Anaconda to Docker quickstart
pick 115ed2215 DOC: fixed Usage:   pip <command> [options]
pick 3694620e8 DOC: revised Docker Quickstart instructions according to review comments
pick e3ab7e476 DOC: matplotlib is only required for building docs
pick 926170351 TST: sparse.linalg: fix arnoldi test seed
pick c045bfe95 BUG: special: use series for `hyp1f1` when it converges rapidly
pick ce59e83db MAINT: special: fix grammar in `hyp1f1` function comments
pick ead61659b TST: special: parametrize `hyp1f1` tests for better error reporting
pick 276a5e2e3 TST: special: continue to inherit from object in test classes
pick efaa30ddc MAINT: special: further improvements to `hyp1f1`
pick 4237176f3 MAINT: special: avoid unnecessary convergence check in `hyp1f1` series
pick 9384e936f TST: special: require Mpmath 1.1.0 for `test_hyperu_around_0`
pick 6cca36f1d FIX: sphinx isattributedescriptor is not available in sphinx < 2.1
pick 309e95b02 MAINT: Update sphinx minimum version in sphinx conf.py
pick 7abd700f6 BUG: stats: Fix the formula for the variance of the noncentral F distribution.
pick d5671a652 DOC: added step 0 to documentation rendering instructions
pick 955cb88a8 ENH: faster circumcenter calculation in spatial.SphericalVoronoi
pick 5ded2b734 multiple stability updates
pick d84045c83 minor editing
pick 9c026fb5a bug fixes for small tolerance - explicit RR
pick 57eaeeafd extended and beautified
pick e671e42cc pep8 fixes
pick c2052a708 added new unit test that crashes the old lobpcg
pick e8af6cb1b typo fix and more matrix types added
pick 95ba7432e documentation updates
pick 38eca4c07 typo in docs/example fixed
pick 102fa0f7d small doc/example edit
pick ab769b9d9 DOC: make lobpcg docstring confirm to numpydoc standard.
pick d867746fe DOC: special: Add documentation for `betainc` function
pick 9f398b948 DOC: special: Update examples, notes in `betainc` docsting
pick e798ba0b6 DOC: special: Add documentation for `beta` function
pick 076896f7a DOC: special: Add "See Also" section to `beta` docstring
pick 04cd360b2 MAINT: remove time print lbfgsb
pick 612a290e7 DOC: updated rendering documentation for circleci-artifacts-redirector
pick dcc7dfd00 DOC: adjusted wording in Contributor Guide TOC
pick 001ad8915 DOC: reinforced need to have proper dev environment active to build docs
pick d68fd9c24 DOC: updated documentation rendering instructions according to gh-10717/gh-10724
```
Mostly I was just hoping to mark some of my commits as fixups, e.g. `115ed221` which was a mistake. Should I do that and leave the rest of the commits alone? Suggestions for what I should have done differently or what I should do now are welcome.

peterbell10 commented:
@mdhaber you could do
```
git rebase master docker_quickstart
```
and it should rebase your branch on master, skipping the merge commits. Then do your interactive rebase as normal.

mdhaber commented:
@peterbell10 hmm that added a lot of commits to the list. I ended up removing everything that I didn't write in the branch and working with that, then force pushed and resolved the merge conflict in the web-based editor. I think this one is ready, but suggestions for the future are appreciated!

peterbell10 commented:
If you used the command I gave verbatim then it could be that your local master is out of date, meaning you would have rebased on an older commit.

mdhaber commented:
I didn't just use the command verbatim. From the `docker_quickstart`, I tried:
```
git fetch upstream
git rebase upstream/master
```
Then I noticed that's not what you had written, so I rebased my local master on upstream/master, _then_ entered your command verbatim.

andyfaff commented:
After forking scipy on github I normally do:

```
git clone https://github.com/andyfaff/scipy.git
# I prefer to use the name of the project as the name
git remote add scipy https://github.com/scipy/scipy.git
git fetch scipy
git checkout <featurebranch>
git rebase scipy/master
```

Then interactively rebase the commits your branch is adding.

andyfaff commented:
I've read through this and it looks mergeable to me. @mdhaber did you want to do anything more with this?

peterbell10 commented:
>@mdhaber you could do `git rebase master`

For future reference, I was actually thinking of `git rebase --onto master` which replays your branches commits onto master. I tried it on a scipy branch after a merge: It did work, but it created a few non-trivial merge conflicts because SciPy's git history contained some fixup commits. It works much better if PRs are usually squash merged.

mdhaber commented:
Thanks for those thoughts, @andyfaff, @peterbell10.

No, I think I'm happy with this as is for now @andyfaff.

andyfaff commented:
Thank you @mdhaber.

mdhaber commented:
Thank you @andyfaff for paving the way! This has been really useful on my Windows machine.
